R/do_svd.R

```diff
@@ -0,0 +1,308 @@
+        # with subjects, new dimentions and values
+        cnames <- avoid_conflict(grouped_col, c(subject_col, "new.dimension", value_cname))
+        rownames(mat) <- rownames(matrix)
+        ret <- mat_to_df(mat, cnames)
+      } else {
+        stop(paste(output, "is not supported as output"))
+      }
+    } else if (type=="dimension") {
+      # v matrix in svd can be regarded as the direction of new axes in original dimension.
+      result <- svd(matrix, nv=n_component, nu=0)
+      mat <- result$v
+
+      rownames(mat) <- colnames(matrix)
+
+      if (output=="wide") {
+        # a column is a base vector of
+        # a new axis of reduced dimension
+        # in the original dimension
+        ret <- as.data.frame(mat)
+        colnames(ret) <- avoid_conflict(c(grouped_col, dimension_col), paste("axis", seq(ncol(mat)), sep=""))
+        rnames <- same_type(rownames(mat), df[[dimension_col]])
+        df <- setNames(data.frame(rnames, stringsAsFactors = FALSE), dimension_col)
+        ret <- cbind(df, ret)
+      } else if (output=="long") {
+        # this returns molten format data frame of matrix
+        # with key, new dimentions and values
+        cnames <- avoid_conflict(grouped_col, c(dimension_col, "new.dimension", value_cname))
+        ret <- mat_to_df(mat, cnames)
+      } else {
+        stop(paste(output, "is not supported as output"))
+      }
+    } else if (type=="variance"){
+      # d vector can be regarded
+      # as how much the data is distributed
+      # in the direction of new axes.
+      variance <- svd(matrix, nu=0, nv=0)$d
+      component <- seq(min(length(variance), n_component))
+      if (output=="wide") {
+        # this return one row data frame with d of svd
+        mat <- matrix(variance[component], ncol=length(component))
+        ret <- as.data.frame(mat)
+        colnames(ret) <- avoid_conflict(c(grouped_col), paste("axis", seq(ncol(mat)), sep=""))
+      } else if (output=="long") {
+        # this return a data frame with a column with d of svd
+        ret <- data.frame(component = component, svd.value = variance[component])
+        colnames(ret) <- avoid_conflict(grouped_col, c("new.dimension", value_cname))
+      } else {
+        stop(paste(output, "is not supported as output"))
+      }
+    } else {
+      stop(paste(type, "is not supported as type argument."))
+    }
+    ret
+  }
+
+  # Calculation is executed in each group.
+  # Storing the result in this tmp_col and
+  # unnesting the result.
+  # If the original data frame is grouped by "tmp",
+  # overwriting it should be avoided,
+  # so avoid_conflict is used here.
+  tmp_col <- avoid_conflict(grouped_col, "tmp")
+  ret <- df %>%
+    dplyr::do_(.dots=setNames(list(~do_svd_each(.)), tmp_col)) %>%
+    unnest_with_drop_(tmp_col)
+
+  ret
+}
+
+#' Calculate distance of each pair of groups.
+#' @param df data frame in tidy format
+#' @param group A column you want to calculate the correlations for.
```

Comment:
The parameters comment does not seem to match with actual parameters. For example, I don't see type here.